File: compiler/bench/treedepth.c
// expected: 17
#include <stdio.h>
long max(long a, long b) { return a > b ? a : b; }

long treedepth(long depth) {
    if (depth == 0) return 0;
    return max(treedepth(depth - 1), treedepth(depth - 1)) + 1;
}

int main() {
    for (int i = 0; i < 1000; i++) {
        treedepth(17);
    }
    printf("%ld\n", treedepth(17));
    return 0;
}

File: compiler/bench/treedepth.fs
\ expected: 17
\ Max depth recursively, simulate 100K node tree

: max ( a b -- max ) 2dup > if drop else nip then ;

: treedepth ( depth -- maxdepth ) recursive
  dup 0= if exit then
  1- dup recurse swap recurse max 1+ ;

: main
  1000 0 do 17 treedepth drop loop
  17 treedepth . cr ;
